src/zjs_ocf_common.c

```diff
@@ -41,178 +55,232 @@ int zjs_ocf_is_int(jerry_value_t val)
-            zjs_rep_end_object(enc, &child);
-            DBG_PRINT("}\n");
-        } else if (jerry_value_is_array(prop)) {
-            /*
-             * TODO: optimization can we recursively call zjs_ocf_props_setup()?
-             */
-            CborEncoder child;
-            DBG_PRINT("Encoding array: %s [\n", h->names_array[i]);
-            zjs_rep_set_array(enc, &child, h->names_array[i]);
-            for (int i = 0; i < (int)jerry_get_array_length(prop); i++) {
-                ZVAL element = jerry_get_property_by_index(prop, i);
-                if (jerry_value_is_number(element)) {
-                    // Number could be double, int, uint
-                    int type = zjs_ocf_is_int(element);
-                    if (type == TYPE_IS_NUMBER) {
-                        double num = jerry_get_number_value(element);
-                        zjs_rep_encode_double(&child, num);
-                        DBG_PRINT("Encoding number: %lf\n", num);
-                    } else if (type == TYPE_IS_UINT) {
-                        unsigned int num = (unsigned int)jerry_get_number_value(element);
-                        zjs_rep_encode_uint(&child, num);
-                        DBG_PRINT("Encoding unsigned int: %u\n", num);
-                    } else if (type == TYPE_IS_INT) {
-                        int num = (int)jerry_get_number_value(element);
-                        zjs_rep_encode_int(&child, num);
-                        DBG_PRINT("Encoding int: %d\n", num);
-                    }
-                } else if (jerry_value_is_boolean(element)) {
-                    bool boolean = jerry_get_boolean_value(element);
-                    zjs_rep_encode_int(&child, boolean);
-                    DBG_PRINT("Encoding boolean: %d\n", boolean);
-                } else if (jerry_value_is_string(element)) {
-                    ZJS_GET_STRING(element, str, OCF_MAX_PROP_NAME_LEN);
-                    zjs_rep_encode_string(&child, str);
-                    DBG_PRINT("Encoding string %s\n", str);
+            break;
+        case BOOL_ARRAY:
+            {
+                sz = oc_bool_array_size(rep->value.array);
+                bool *b_array = oc_bool_array(rep->value.array);
+                jerry_value_t array = jerry_create_array(sz);
+                for (i = 0; i < sz; i++) {
+                    jerry_value_t b = jerry_create_number(b_array[i]);
+                    jerry_set_property_by_index(array, i, b);
+                    jerry_release_value(b);
                 }
+                zjs_set_property(object, oc_string(rep->name), array);
+                jerry_release_value(array);
             }
-            // Close the array
-            zjs_rep_close_array(enc, &child);
-            DBG_PRINT("]\n");
+            break;
+        case OBJECT:
+            {
+                jerry_value_t new_object = zjs_create_object();
+                decode_value_helper(new_object, rep->value.object);
```

Comment:
decode_value_helper() returns a object?  Could this be leak here if you don't free it?

Reply:
It returns ZJS_UNDEFINED *most* of the time, but in the case of an object nested inside the root object yes it will return a new object. This if statement checks if the object is not named, if it is indeed not named then its part of an array, meaning its iterating the OBJECT_ARRAY case below. In that case the return value gets freed.

Unfortunately having this special case was a bit messy, but the only way I could think of to implement the OBJECT_ARRAY type.

Reply:
ok, since i just see that returned obj gets released down further, just checking